ledger-service/http-json/src/itlib/scala/http/AbstractWebsocketServiceIntegrationTest.scala

```diff
@@ -916,6 +953,173 @@ abstract class AbstractWebsocketServiceIntegrationTest
       }
   }
 
+  "Per-query offsets should work as expected" in withHttpService { (uri, _, _) =>
+    val dslSyntax = Consume.syntax[JsValue]
+    import dslSyntax._
+    import spray.json._
+    def createIouCommand(currency: String): String =
+      s"""{
+           |  "templateId": "Iou:Iou",
+           |  "payload": {
+           |    "observers": [],
+           |    "issuer": "Alice",
+           |    "amount": "999.99",
+           |    "currency": "$currency",
+           |    "owner": "Alice"
+           |  }
+           |}""".stripMargin
+    def createIou(currency: String): Future[Assertion] =
+      HttpServiceTestFixture
+        .postJsonStringRequest(
+          uri.withPath(Uri.Path("/v1/create")),
+          createIouCommand(currency),
+          headersWithAuth,
+        )
+        .map(_._1 shouldBe a[StatusCodes.Success])
+    def contractsQuery(currency: String): String =
+      s"""{"templateIds":["Iou:Iou"], "query":{"currency":"$currency"}}"""
+    def contractsQueryWithOffset(offset: domain.Offset, currency: String): String =
+      s"""{"templateIds":["Iou:Iou"], "query":{"currency":"$currency"}, "query_offset":"${offset.unwrap}"}"""
+    def contracts(currency: String, offset: Option[domain.Offset]): String =
+      offset.fold(contractsQuery(currency))(contractsQueryWithOffset(_, currency))
+    def ledgerEnd(expectedContracts: Int): Future[domain.Offset] = {
+      def go(killSwitch: UniqueKillSwitch): Sink[JsValue, Future[domain.Offset]] =
+        Consume.interpret(
+          for {
+            _ <- readAcsN(expectedContracts)
+            ContractDelta(Vector(), _, Some(offset)) <- readOne
+            _ = killSwitch.shutdown()
+            _ <- drain
+          } yield offset
+        )
+      val (killSwitch, source) =
```

Comment:
you are probably paying some overhead for establishing separate streams for each call to `getLedgerEnd`. What I’d probably try is to hardcode the offsets temporarily and see if `getLedgerEnd` plays a significant role in the slowdown or note.

Reply:
I was looking at that and measured the gap between retrieving the ACS and getting the ledger end and it's consistently ~300 micros.